narwhal/primary/src/synchronizer.rs

```diff
@@ -74,6 +344,272 @@ impl Synchronizer {
+                digest, e
+            );
+            return Err(DagError::ShuttingDown);
+        }
+
+        if let Some(sender) = pending.remove(&digest) {
+            // Ignore error if there is no more listeners.
+            let _ = sender.send(());
+        }
+
+        Ok(())
+    }
+
+    // TODO: allow multiple (<=10) outgoing requests concurrently, and multiple certificates
+    // per request.
+    async fn push_certificates(
+        network: Network,
+        name: BLS12381PublicKey,
+        network_key: NetworkPublicKey,
+        mut rx_own_certificate_broadcast: broadcast::Receiver<Certificate>,
+    ) {
+        const PUSH_TIMEOUT: Duration = Duration::from_secs(5);
+        const MIN_FAILURE_BACKOFF: Duration = Duration::from_millis(100);
+        const MAX_FAILURE_BACKOFF: Duration = Duration::from_secs(10);
+        let peer_id = anemo::PeerId(network_key.0.to_bytes());
+        let peer = network.waiting_peer(peer_id);
+        let mut client = PrimaryToPrimaryClient::new(peer);
+        let mut certificates = VecDeque::new();
+        let mut failure_backoff = MIN_FAILURE_BACKOFF;
+        loop {
+            let wait_timeout = if certificates.is_empty() {
+                MAX_FAILURE_BACKOFF
+            } else {
+                failure_backoff
+            };
+            match timeout(wait_timeout, rx_own_certificate_broadcast.recv()).await {
+                Ok(Ok(cert)) => certificates.push_back(cert),
+                Ok(Err(broadcast::error::RecvError::Closed)) => {
+                    trace!("Certificate sender {name} shutting down!");
+                    return;
+                }
+                Ok(Err(broadcast::error::RecvError::Lagged(e))) => {
+                    warn!("Certificate broadcaster {name} lagging! {e}");
+                    // Re-run the loop to receive again.
+                    continue;
+                }
+                Err(_) => {
+                    if certificates.is_empty() {
+                        // MAX_FAILURE_BACKOFF has elapsed without a certificate created.
+                        // This should rarely happen.
+                        continue;
+                    }
+                }
+            };
+            // Get more certificates if available in the broadcast channel.
+            loop {
+                match rx_own_certificate_broadcast.try_recv() {
+                    Ok(cert) => {
+                        certificates.push_back(cert);
+                    }
+                    Err(broadcast::error::TryRecvError::Closed) => {
+                        trace!("Certificate sender {name} shutting down!");
+                        return;
+                    }
+                    Err(broadcast::error::TryRecvError::Lagged(e)) => {
+                        warn!("Certificate sender {name} lagging! {e}");
+                    }
+                    Err(broadcast::error::TryRecvError::Empty) => {
+                        break;
```

Comment:
As we do have nested loops in this code I would suggest [naming them](https://doc.rust-lang.org/rust-by-example/flow_control/loop/nested.html) to make the intention super clear.

Reply:
Sure. Also this logic will be refactored soon to allow multiple certificates inflight being sent.